### DISCUSSION TITLE
Implement my own VirtualizingWrapPanel control outside of avalonia

### DISCUSSION THREAD

**Discussion by Whiletru3** [AvaloniaUI/Avalonia#15445] [Q&A] [answered]
Hi,

I try to implement my own control VirtualizingWrapPanel  (based on this PR https://github.com/AvaloniaUI/Avalonia/pull/11251).
Because this control has limitations (like every item has to be size equal), putting it directly in avalonia is excluded.

But for now, I can't use it outside of avalonia.Controls because of the accessibility of some methods .
In VirtualizingPanel :
protected internal override IEnumerable<Control>? GetRealizedContainers()
protected internal override Control? ContainerFromIndex(int index)
protected internal override int IndexFromContainer(Control container) => _realizedElements?.GetIndex(container) ?? -1;
protected internal override Control? ScrollIntoView(int index)

In ILayoutRoot : 
internal ILayoutManager LayoutManager { get; }
In TopLevel :
internal ILayoutManager LayoutManager

Is it possible to change it as public instead of internal ?
Is there a more elegant way ? 

Thanks

**Comment by BAndysc** [comment]
> But for now, I can't use it outside of avalonia.Controls because of the accessibility of some methods .

`protected internal` is a weird visibility modifier in c#, but despite its confusing name - you can use it in your derived class (not only can, but even must, because those methods are abstract in VirtualizingPanel).

As for `LayoutManager` - it was used only in tests, do you need to port the tests as well?

**Reply by Whiletru3** [reply]
I can't use it outside of avalonia in my own software because it is internal :-/

**Reply by BAndysc** [reply]
You **can** use it outside of Avalonia, because "protected internal" does NOT mean "protected AND internal", it means "protected OR internal", meaning in Avalonia this method can be called from any class in assembly (internal), whereas when you use it outside of Avalonia, you can call it only from within the class and derived classes (protected). Of course you don't call this method, only override which you can also do. https://stackoverflow.com/a/51241984/1616645

**Reply by Whiletru3** [reply]
Thanks, you're right, ok for those if I remove the internal in my override : 
protected internal override IEnumerable? GetRealizedContainers()
protected internal override Control? ContainerFromIndex(int index)
protected internal override int IndexFromContainer(Control container) => _realizedElements?.GetIndex(container) ?? -1;
protected internal override Control? ScrollIntoView(int index)

But for LayoutManager it seems that it is used on purpose and not for Tests :
```
// If the item being brought into view was added since the last layout pass then
// our bounds won't be updated, so any containing scroll viewers will not have an
// updated extent. Do a layout pass to ensure that the containing scroll viewers
// will be able to scroll the new item into view.
if (!Bounds.Contains(rect) && !_viewport.Contains(rect))
{
    _isWaitingForViewportUpdate = true;
    root.LayoutManager.ExecuteLayoutPass();
    _isWaitingForViewportUpdate = false;
}
```

I checked, in Avalonia.Controls, there are InternalVisibleTo for other assemblies (like tests of ItemsRepeater).

So for this one I'm stucked :/

**Reply by Whiletru3** [reply]
Ho yes, I found the replacing feature in the commit description 👍

- ILayoutManager is hidden
- - LayoutManager.ExecuteLayoutPass() exposed as TopLevel.UpdateLayout()

**Comment by maxkatz6** [comment]
There is Avalonia.Labs branch with a basic VirtualizingWrapPanel implementation: https://github.com/AvaloniaUI/Avalonia.Labs/tree/virtualizing_wrappanel

But VirtualizingWrapPanel is nearly impossible to get right. VirtualizingUniformGrid though is possible.

**Reply by Whiletru3** [reply]
Hi @maxkatz6,

This Avalonia.Lab branch is facing the same issue as mine for the LayoutManager :( (see this comment : https://github.com/AvaloniaUI/Avalonia/discussions/15445#discussioncomment-9187311 )
And since last year, the VirtualizingStackPanel got a lot of corrections/improvements. I try to take all those in a newer version.
But the real showstopper here is the accessibility of the LayoutManager.

**Reply by Whiletru3** [reply, marked as answer]
Ho yes, I found the replacing feature in the commit description 👍

- ILayoutManager is hidden
- - LayoutManager.ExecuteLayoutPass() exposed as TopLevel.UpdateLayout()
